Lib/zipimport.py

```diff
@@ -305,7 +305,18 @@ def __repr__(self):
 # Given a module name, return the potential file path in the
 # archive (without extension).
 def _get_module_path(self, fullname):
-    return self.prefix + fullname.rpartition('.')[2]
+    path = self.prefix + fullname.rpartition(".")[2]
+
+    # Here be dragons.
+    # Originally, this *only* used rpartition('.')[2] to find
+    # the last item in a namespace package, but that made it basically
+    # impossible to use nested submodules.
+    for suffix, _, _ in [*_zip_searchorder, (path_sep, False, False)]:
+        if (path + suffix) in self._get_files():
+            # OK, we found it the old way. Let's return it.
+            return path
+
+    return self.prefix + fullname.replace(".", path_sep)
```

Comment:
As https://github.com/python/cpython/pull/121161/files#r1660062444 described, this patch has no effect, shoud patch `_is_dir` or `_read_directory`.

Reply:
> As https://github.com/python/cpython/pull/121161/files#r1660062444 described, this patch has no effect, shoud patch `_is_dir` or `_read_directory`.

From what I found, there's no problem with `_read_directory`, and `_is_dir` is only relevant to `find_spec` (the problem still occurs in `get_code`, for example). This patch seems to fix the issue for both. Have you tried this fix on [this issue](https://github.com/SeaHOH/memimport/issues/3)?

Reply:
For namespace package's dealing and operation, there has already clearly explained in the source code comment. You can see them in `find_loader` (<=py311), `find_spec` and `_is_dir`.

And, why your test with `get_code` (and `get_filename`, `get_source`, `is_package` also) was fail? Because find functons depend the loader's path prefix, different level modules have different level loaders, skip levels will be failed. Yes, this is `_get_module_path` which you patched, then it broken, return the full path at once instead of stepwise lookup.

If you like to test into zipimport functions, it should be `get_data`:
```py
    def testImportSubmodulesInZip(self):
        with ZipFile(TEMP_ZIP, "w") as z:
            z.writestr("a/__init__.py", b'')
            z.writestr("a/b/c/__init__.py", b'def foo(): return "foo"')

        importer = zipimport.zipimporter(TEMP_ZIP + "/a/")
        spec = importer.find_spec("a.b")
        self.assertEqual(spec.loader, None)
        self.assertEqual(spec.submodule_search_locations,
                         [(TEMP_ZIP + "/a/b/").replace("/", zipimport.path_sep)])
        self.assertRaises(zipimport.ZipImportError, importer.get_code("a.b"))
        self.assertEqual(importer.get_data("a/b/"), b"")

        sys.path.insert(0, TEMP_ZIP)
        try:
            import a.b.c
        else:
            self.assertIn('namespace', str(a.b))
            self.assertEqual(a.b.c.foo(), "foo")
        finally:
            del sys.path[0]
            sys.modules.pop('a.b.c', None)
            sys.modules.pop('a.b', None)
            sys.modules.pop('a', None)
```

Reply:
I agree with @SeaHOH that the problem is rather in `_read_directory()`. I created #121233 for this approach.